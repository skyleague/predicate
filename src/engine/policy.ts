import { collect, stack } from '@skyleague/axioms'
import type { IsEmptyObject, Simplify } from '@skyleague/axioms/types'
import packageJSON from '../../package.json' with { type: 'json' }
import type { FactsFomExprs } from './operator.js'
import type { Expression, ExpressionReturnType } from './types.js'

const { version } = packageJSON

export class EvaluationContext {
    public evaluated = new WeakMap<object, unknown>()
    public state: Partial<Record<PropertyKey, unknown>> = {}
    public input: Record<string, unknown> = {}
    public scope: unknown = {}
    public stack: (readonly [EvaluationContext['evaluated'], EvaluationContext['state'], unknown])[]

    public constructor(input: Record<string, unknown>) {
        this.input = input
        this.stack = [[this.evaluated, this.state, this.scope]]
    }

    public evaluate<O>(expr: Expression<O>): O {
        if (this.evaluated.has(expr)) {
            return this.evaluated.get(expr) as O
        }

        const input = expr.dependsOn.map((d) => this.evaluate(d))

        const result = expr.fn(input, this)
        this.evaluated.set(expr, result)
        return result
    }

    public withScope<T>(scope: unknown, fn: () => T) {
        this.push(scope)
        const result = fn()
        this.pop()
        return result
    }

    protected push(scope: unknown) {
        const top = [new WeakMap<object, unknown>(), {}, scope] as const
        this.stack.push(top)
        this.evaluated = top[0]
        this.state = top[1]
        this.scope = top[2]
    }

    protected pop() {
        this.stack.pop()
        const top = this.stack.at(-1) ?? [this.evaluated, this.state, this.scope]
        this.evaluated = top[0]
        this.state = top[1]
        this.scope = top[2]
    }
}

export interface _EvaluationContext {
    evaluated: WeakMap<object, unknown>
    state: Partial<Record<PropertyKey, unknown>>
    stack: [_EvaluationContext['evaluated'], _EvaluationContext['state'], unknown][]
    scope: unknown
    evaluate<T>(expr: Expression<T>): T
    withScope: <T>(y: unknown, fn: () => T) => T
}

function* collapseExpression(root: Expression[], seen = new WeakSet()) {
    const nodes = stack(root)
    for (const n of nodes) {
        if (!seen.has(n)) {
            yield n
            nodes.push(n.dependsOn)
            seen.add(n)
        }
    }
}
type InferFactName<Expr, k> = Expr extends { name: string } ? Expr['name'] : k

type FilterFactExpressions<Facts, K extends keyof Facts = keyof Facts> = Facts[K] extends {
    _type: 'fact'
}
    ? InferFactName<Facts[K], K>
    : never

type _InputFromName<Names extends PropertyKey, Facts> = {
    [K in Names]: ExpressionReturnType<Extract<Facts[keyof Facts], { name: K }>>
}
type _InputFromExpressions<Facts> = _InputFromName<
    FilterFactExpressions<Facts> extends PropertyKey ? FilterFactExpressions<Facts> : never,
    Facts
>

type _FactsFomExprs<Facts> = Facts extends unknown[] ? Facts[number] : Facts
export type InputFromExpressions<Facts extends Record<string, unknown>> = Simplify<
    _InputFromExpressions<{ [K in keyof Facts as FilterFactExpressions<Facts, K>]: Facts[K] }> &
        _InputFromExpressions<{ [K in keyof Facts]: _FactsFomExprs<FactsFomExprs<Facts[K]>> }>
>

export type OutputFromFacts<Facts extends Record<string, unknown>> = Simplify<{
    [k in keyof Facts]: ExpressionReturnType<Facts[k]>
}>

export interface Policy<I, O> {
    evaluate: IsEmptyObject<I> extends true ? () => { input: I; output: O } : (x: I) => { input: I; output: O }
    expr: () => unknown
}

export function $policy<Facts extends Record<string, Expression>>(
    expressions: Facts,
): Policy<InputFromExpressions<Facts>, OutputFromFacts<Facts>> {
    const facts = Object.entries(expressions).map(([name, e]) => {
        if (e.name === undefined) {
            return Object.assign(e, { name })
        }
        return e
    })

    const allNodes = collect(collapseExpression(facts))
    const inputNodes = allNodes.filter(
        (f: (typeof allNodes)[number]): f is (typeof allNodes)[number] & { name: string } =>
            f._type === 'fact' && f.name !== undefined,
    )
    const outputNodes = allNodes.filter((f) => f._type !== 'fact' && f.name !== undefined)

    const properties = Object.fromEntries(inputNodes.map((e) => [e.name, e.expr('definition')]))
    const outputExpression = Object.fromEntries(outputNodes.map((f) => [f.name, f.expr('expression')]))
    return {
        evaluate: ((input: Record<string, unknown>) => {
            const ctx = new EvaluationContext(input)

            for (const fact of facts) {
                const result = ctx.evaluate(fact)
                if (fact.name !== undefined) {
                    ctx.state[fact.name] = result
                }
            }

            return { input: input, output: ctx.state }
        }) as Policy<InputFromExpressions<Facts>, OutputFromFacts<Facts>>['evaluate'],
        expr: () => {
            const input = properties
            return {
                meta: {
                    version,
                },
                input,
                output: outputExpression,
            }
        },
    }
}
