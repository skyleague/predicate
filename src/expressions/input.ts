import type {
    FactExpression,
    DefinitionType,
    Expression,
    InputExpression,
    InferExpressionType,
    LiteralExpression,
} from '../engine/types.js'
import type { FromExpr } from '../json/jsonexpr.type.js'

import { type JSONPathValue, JSONPath } from '@skyleague/jsonpath'
import type { Schema } from '@skyleague/therefore'

import { inspect } from 'node:util'

export interface Fact<T, Name extends string> extends FactExpression<T> {
    name: Name
    dependsOn: []
    _type: 'fact'
}

export function $fact<T, Name extends string>(schema: Pick<Schema<T>, 'schema' | 'is'>, name: Name): Fact<T, Name> {
    return {
        _type: 'fact',
        dependsOn: [],
        name: name,
        fn: ((_: unknown, ctx: { input: Record<Name, T> }) => {
            return ctx.input[name]
        }) as Fact<T, Name>['fn'],
        expr: (definition: DefinitionType) =>
            (definition === 'definition' ? schema.schema : { fact: name.toString() }) as InferExpressionType<T>,
        [inspect.custom]() {
            return `$fact({schema: ${JSON.stringify(schema.schema)}}, "${name}")`
        },
    }
}
export interface From<I, O, DependsOn extends Expression[]> extends InputExpression<O, I, DependsOn, FromExpr> {
    _type: 'value'
}

export function $from<T, Name extends string>(fact: Fact<T, Name>): From<T, JSONPathValue<T, '$'>, [Fact<T, Name>]>
export function $from<T, P extends string, Name extends string>(
    fact: Fact<T, Name>,
    path: P
): From<T, JSONPathValue<T, P>, [Fact<T, Name>]>
export function $from<T, P extends string, Name extends string>(
    fact: Fact<T, Name>,
    path: P = '$' as P
): From<T, JSONPathValue<T, P>, [Fact<T, Name>]> {
    return {
        _type: 'value',
        dependsOn: [fact],
        fn: ((_: unknown, ctx: { input: Record<Name, T> }) => JSONPath.get(ctx.input[fact.name], path)) as From<
            T,
            JSONPathValue<T, P>,
            [Fact<T, Name>]
        >['fn'],
        expr: (_definition) => ({ from: [fact.name, path.toString()] as const }),
        [inspect.custom]() {
            return `$from(${fact[inspect.custom]?.() ?? `"${fact.name}"`}${path === '$' ? '' : `, "${path}"`})`
        },
    }
}

export function $literal<O>(x: O): LiteralExpression<O> {
    return {
        fn: () => x,
        dependsOn: [],
        expr: () => x as InferExpressionType<O>,
        _type: 'literal',
        [inspect.custom]() {
            return `$literal(${inspect(x)})`
        },
    }
}
