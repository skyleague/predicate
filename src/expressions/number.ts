import { operator } from '../engine/operator.js'

export const $add = operator({ operator: '+', symbol: '$add', fn: ([a, b]: [number, number]) => a + b })
export const $subtract = operator({ operator: '-', symbol: '$subtract', fn: ([a, b]: [number, number]) => a - b })
export const $divide = operator({ operator: '/', symbol: '$divide', fn: ([a, b]: [number, number]) => a / b })
export const $multiply = operator({ operator: '*', symbol: '$multiply', fn: ([a, b]: [number, number]) => a * b })
export const $modulo = operator({ operator: '%', symbol: '$modulo', fn: ([a, b]: [number, number]) => a % b })
