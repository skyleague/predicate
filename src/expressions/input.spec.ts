import { forAll, tuple } from '@skyleague/axioms'
import { arbitrary } from '@skyleague/therefore'
import { describe, expect, it } from 'vitest'
import { z } from 'zod'
import { Arithmetic } from '../../test/arithmatic.type.js'
import { Person } from '../../test/person.type.js'
import { $policy } from '../engine/policy.js'
import { $fact, $from } from './input.js'
import { $add } from './number.js'

describe('fact', () => {
    it('simple single one is used as input', () => {
        const person = $fact(Person, 'person')
        const policy = $policy({ person })
        forAll(arbitrary(Person), (p) => {
            expect(policy.evaluate({ person: p }).input.person).toEqual(p)
        })

        expect(policy.expr()).toMatchInlineSnapshot(`
        {
          "input": {
            "person": {
              "$schema": "http://json-schema.org/draft-07/schema#",
              "additionalProperties": true,
              "properties": {
                "age": {
                  "type": "number",
                },
                "birthDate": {
                  "type": "string",
                },
                "firstName": {
                  "type": "string",
                },
                "lastName": {
                  "type": "string",
                },
              },
              "required": [
                "firstName",
                "lastName",
                "birthDate",
                "age",
              ],
              "title": "Person",
              "type": "object",
            },
          },
          "meta": {
            "version": "1.0.0",
          },
          "output": {},
        }
      `)
    })

    it('simple single one is used as input - zod', () => {
        const zodPerson = z.object({
            firstName: z.string(),
            lastName: z.string(),
            birthDate: z.string(),
            age: z.number(),
        })
        zodPerson.safeParse
        const person = $fact(zodPerson, 'person')
        const policy = $policy({ person })
        forAll(arbitrary(Person), (p) => {
            expect(policy.evaluate({ person: p }).input.person).toEqual(p)
        })

        expect(() => policy.expr()).toThrowErrorMatchingInlineSnapshot(
            '[Error: Cannot infer definition for fact with zod schema]',
        )
    })

    it('simple multiple is used as input', () => {
        const person = $fact(Person, 'person')
        const arithmetic = $fact(Arithmetic, 'arith')
        const a = $from(arithmetic, '$.a')
        const b = $from(arithmetic, '$.b')
        const ab = $add(a, b)

        const policy = $policy({ ab, person })

        forAll(tuple(arbitrary(Person), arbitrary(Arithmetic)), ([p, arith]) => {
            const evaluation = policy.evaluate({ person: p, arith: arith })
            expect(evaluation.input.person).toEqual(p)
            expect(evaluation.input.arith).toEqual(arith)
        })

        expect(policy.expr()).toMatchInlineSnapshot(`
          {
            "input": {
              "arith": {
                "$schema": "http://json-schema.org/draft-07/schema#",
                "additionalProperties": true,
                "properties": {
                  "a": {
                    "type": "number",
                  },
                  "b": {
                    "type": "number",
                  },
                },
                "required": [
                  "a",
                  "b",
                ],
                "title": "Arithmetic",
                "type": "object",
              },
              "person": {
                "$schema": "http://json-schema.org/draft-07/schema#",
                "additionalProperties": true,
                "properties": {
                  "age": {
                    "type": "number",
                  },
                  "birthDate": {
                    "type": "string",
                  },
                  "firstName": {
                    "type": "string",
                  },
                  "lastName": {
                    "type": "string",
                  },
                },
                "required": [
                  "firstName",
                  "lastName",
                  "birthDate",
                  "age",
                ],
                "title": "Person",
                "type": "object",
              },
            },
            "meta": {
              "version": "1.0.0",
            },
            "output": {
              "ab": {
                "+": [
                  {
                    "from": [
                      "arith",
                      "$.a",
                    ],
                  },
                  {
                    "from": [
                      "arith",
                      "$.b",
                    ],
                  },
                ],
              },
            },
          }
        `)
    })
})

describe('from', () => {
    it('simple single one is used as input', () => {
        const person = $fact(Person, 'person')
        const firstName = $from(person, '$.firstName')
        const policy = $policy({ person, firstName })
        forAll(arbitrary(Person), (p) => {
            expect(policy.evaluate({ person: p }).input.person).toEqual(p)
        })

        expect(policy.expr()).toMatchSnapshot()
    })
})
