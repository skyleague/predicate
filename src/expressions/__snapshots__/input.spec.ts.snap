// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`from > simple single one is used as input 1`] = `
{
  "input": {
    "person": {
      "$schema": "http://json-schema.org/draft-07/schema#",
      "additionalProperties": true,
      "properties": {
        "age": {
          "type": "number",
        },
        "birthDate": {
          "type": "string",
        },
        "firstName": {
          "type": "string",
        },
        "lastName": {
          "type": "string",
        },
      },
      "required": [
        "firstName",
        "lastName",
        "birthDate",
        "age",
      ],
      "title": "Person",
      "type": "object",
    },
  },
  "meta": {
    "version": "1.0.0",
  },
  "output": {
    "firstName": {
      "from": [
        "person",
        "$.firstName",
      ],
    },
  },
}
`;
